usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -2709,6 +3096,26 @@ CK_RV token_specific_object_add(STDLL_TokData_t * tokdata, SESSION * sess,
         return CKR_OK;
     }
 
+    /* Check that we do not try to create an attribute bound private key. */
+    /* need class for private/public key info */
+    rc = template_attribute_get_ulong(obj->template, CKA_CLASS,
+                                      &class);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Could not find CKA_CLASS for the key.\n");
+        return rc;
+    }
+    rc = template_attribute_get_bool(obj->template, CKA_IBM_ATTRBOUND,
+                                    &attrbound);
+    if (rc == CKR_TEMPLATE_INCOMPLETE) {
+        attrbound = false;
+    } else if (rc != CKR_OK) {
+        TRACE_ERROR("Incomplete CKA_IBM_ATTRBOUND attribute for the key.\n");
+        return rc;
+    }
+    if (class != CKO_PUBLIC_KEY && attrbound) {
```

Comment:
How about CKO_SECRET_KEY? 
Is an attr bound public key the only key that can be created via C_CreateObject?

Reply:
Yes, since it does not contain any secrets ;-)  Everything that contains secrets and should be AB needs special treatment since the secrets are store in encrypted for in AB (which we currently do not support for C_CreateObject).  This is also essential to really separate the key hierarchies.